### DISCUSSION TITLE
remark-directive compatibility with emoji plugins

### DISCUSSION THREAD

**Discussion by baumandm** [remarkjs/remark#651] [Ideas] [answered]
I'd like to use the [remark-directive](https://github.com/remarkjs/remark-directive) plugin alongside an `:emoji:` shortcode plugin like [remark-gemoji](https://github.com/remarkjs/remark-gemoji) or [remark-emoji](https://github.com/rhysd/remark-emoji). 

The `:emoji:` shortcode syntax is ubiquitous thanks to GitHub, Slack, etc., so it's a must-have for me.  And the generic directive syntax seems like the most popular solution for adding leaf & block-level elements.

## The Problem

The problem is that the `remark-directive` plugin adds new syntax to the parser that creates `textDirective` nodes.  So a document like this:

```
:emoji:
```

Will be parsed into a tree like

```
{ 
  type: 'paragraph', 
  children: [
    { 
      type: 'textDirective', 
      name: 'emoji', 
      attributes: [Object], 
      children: [Array], 
      position: [Object]
    },
    { 
      type: 'text', 
      value: ':', 
      position: [Object]
    }
  ]
}
```

This breaks emoji plugins that look for `:emoji:` shortcodes in `text` nodes.

## Idea A: Backtrack if there's a trailing colon

Based on the proposed specification and the current implementation, a trailing colon isn't allowed in the text directive (this is why it appears as a separate `text` node).  So one solution would be to update the tokenizer in [micromark-extension-directive](https://github.com/micromark/micromark-extension-directive) to check for a trailing colon after the name.  If found, it would backtrack and fallback to using a `text` node for the entire shortcode.

One major issue with this idea is that currently, there is no spacing requirements to the text directives. For example:

```
:a:b
```

This creates two `textDirective` nodes, `a` and `b`.  The above proposal would turn that into a single `text` node with `:a:b`.  I don't have any ideas on how to avoid this breaking change.

## Idea B: Emojis as text directives

I also considered the feasibility of creating a directive-compatible emoji plugin (e.g. `remark-directive-emoji`) which expects the emoji shortcodes to be converted to `textDirective` nodes.  This plugin could visit `textDirective` nodes and convert them (if needed) into emojis.  

However, with the current parser, the trailing colon is dangling in another `text` node and cleaning it up would be quite hacky.  Alternately it could be parsed into the `textDirective` name, but that's also a victim of the same ambiguity as mentioned above.  

There's also the issue that a `textDirective` node could be both an emoji shortcode and another directive.  This could probably be solved by ensuring the plugins are in the correct priority order.

Any other ideas to get the syntax for both plugins to work together?

**Comment by wooorm** [comment, marked as answer]
Hmm. Interesting Q. And thanks for such a detailed write up. You’ve answered most of the questions I would’ve otherwise asked!

Here is the big complexity: `:\-1:` yields :\-1:.

What what means is that the gemoji shortcodes run *after* markdown is parsed. It’s not a “construct”. The escapes have been used already. So directives and shortcodes run at different stages...

**Reply by wooorm** [reply]
One solution might be, but I’m not 100% if I like it, to have a list of “allowed” directives. E.g., if parsing at `:smile:` knows that `:smile` is not a directive, it doesn’t parse it as such, and instead leaves it as normal text...

**Reply by baumandm** [reply]
This raises another option: creating a new `micromark-extension-emoji` plugin that parses out and generates `shortcode` nodes, in contrast to the existing emoji plugins that operate on top of `text` nodes.  Then downstream plugins could operate on those `shortcode` nodes however they like.

Would that make sense?

**Reply by baumandm** [reply]
> One solution might be, but I’m not 100% if I like it, to have a list of “allowed” directives. E.g., if parsing at `:smile:` knows that `:smile` is not a directive, it doesn’t parse it as such, and instead leaves it as normal text...

That would work for my use cases.  

It also would avoid the situation where I need to provide some kind of renderer for unexpected directives.

**Reply by wooorm** [reply]
- I don’t like Idea A (in the OP), because I think `:strong[Title]: Definition` should work, and as the brackets are optional, then `:strong: Definition` should also “work” IMO. (different example `:nobr{.class}:` and `:nobr:` or so)
- You “debunked” Idea B yourself nicely
- New idea A (`micromark-extension-emoji`) *could* be done, but it’d need to know about escapes too, and I *think* it wouldn’t match GH 100%, due to running at a different stage as I explained above, but I’m failing to come up with a case
- My idea A (“allowed”) still feels weird to me. I think directives are supposed to be “generic”, in that their syntax will work at the parsing stage even though there might not be a handler for them later on.

Complex!

Workaround: you should be able to do `\:smile:`. It can’t be a directive because it’s escaped. But GH will see the `:smile:` when handling the tree to figure out emoji: `\:smile:` -> \:smile:

**Reply by baumandm** [reply]
Yeah there's not a clear winner here IMHO.

Regarding your first bullet point, I think it's a question of precedence.  If there's an emoji `:strong:` (💪) and a directive `:strong` one of them will need to take precedence, but either one would be equally valid.

Another idea for your consideration: throughout the [directive discussion thread](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444), people suggested alternate variations of the syntax, which range from very different to merely swapping the `:` for a `!` or `@` character.  What if the unused [`syntaxOptions`](https://github.com/micromark/micromark-extension-directive#syntaxoptions) map allowed changing the default prefix for inline/leaf/block directives?

Something like:

```
var result = micromark(doc, {
  extensions: [syntax({ leafPrefix: '@', inlinePrefix: '@@', blockPrefix: '@@@' })],
  htmlExtensions: [html({abbr: abbr})]
})
```

(and bubbled up as an option in `remark-directive` as well)

I believe just changing the leaf prefix would be enough to avoid conflicts with the emoji syntax.  It would also make the directive plugin more appealing for people who favor another syntax for other reasons as well.

**Reply by wooorm** [reply]
So changing the marker be a suitable solution for you? I’m all for it. There’s no spec, so I want to stick as close with what’s defined there. But I’m definitely open to options.

I’d prefer using a single `character` option though!

**Reply by baumandm** [reply]
Yea that would work for me.  As you mentioned there's no spec, so I don't have to worry about breaking compatibility with anything else, and my users will likely have never encountered this generic directive proposal before.

The reason I proposed splitting up the prefix options would be to allow more flexibility.  I was specifically thinking about using a single slash-command style for leaf directives:`/leaf`, which would feel more familiar with many of my users.

But I understand that would complicate the existing parser quite a bit, so I'm fine with a single character option.

**Reply by wooorm** [reply]
I was thinking about this some more today and this would also affect serializing (https://github.com/syntax-tree/mdast-util-directive).

> As you mentioned there's no spec, so I don't have to worry about breaking compatibility with anything else, and my users will likely have never encountered this generic directive proposal before.

This is a concern for me. The goal with this syntax is to have it work in different places. Of course, it’s not completely specced. But one of the goals here is for users of my site to use the same syntax on your site, and not have to learn a new syntax, even though there are different “components” 🤔

**Reply by baumandm** [reply]
I agree, it would be preferable if the syntax for both could coexist without issue.  But if that's not possible, there's substantially more precedent for the emoji syntax which is why I am much more willing to adjust the directive syntax.

Perhaps it's worth revisiting idea A from my original post.  Other inline Markdown elements also require spaced separation from other elements, for example `a``b` and *a**b*:

```md
`a``b`
*a**b*
```

With that in mind, it seems consistent if two inline directives needed to be spaced apart to parse correctly.  Additionally, adding an empty `[]` or `{}` could force it to be parsed as a directive.  

Examples:
- `:strong:` - Emoji
- `:strong :` - Directive, text (includes space)
- `:strong{}:` - Directive, text (no space)
- `:strong :nobr` - Directive, directive (includes space)
- `:strong{}:nobr` - Directive, directive (no space)

Thoughts?

**Reply by wooorm** [reply]
That’s a great idea!

**Reply by baumandm** [reply]
I've tested out the change (https://github.com/micromark/micromark-extension-directive/pull/9) and it's working perfectly!

Thank you for all the help.

**Reply by farnabaz** [reply]
The issue still exists in emojis with `_`. like `: call_me_hand:` and `: thinking_face:`

**Reply by wooorm** [reply]
can you make a reproduction?

**Reply by farnabaz** [reply]
Thanks to fast response @wooorm 

Clone the repo and start the project, `:heart:` emoji is there but  `:call_me_hand:`  is wrong
https://github.com/farnabaz/remark-directive-emoji

**Reply by wooorm** [reply]
Thanks!

 So, say we’re using `heart_eyes`, `remark-directive` will see it as `:heart` (a directive) and then leave the rest (`_eyes:`)...
 
We *could* allow `_` in directive names? That would run to the end, `:heart_eyes|` and then see the `:` there?

Or we could disallow the underscore at the “end” of a directive?

@baumandm what do you prefer?

**Reply by baumandm** [reply]
I think both allowing `_` in directive names but not at the end of the directive makes sense.  I think that would avoid issues with emphasis like `_:directive_`?

**Reply by wooorm** [reply]
underscore emphasis is already not allowed in words by commonmark, but that’s smart, should be checked indeed

**Reply by wooorm** [reply]
PR: https://github.com/micromark/micromark-extension-directive/pull/10

**Reply by wooorm** [reply]
landed: https://github.com/micromark/micromark-extension-directive/releases/tag/1.4.0
